lib/cn/csched_sp3_work.c

```diff
@@ -471,30 +471,37 @@ sp3_work_wtype_length(
             return kvsets;
         }
 
-        /* Don't let lightweight nodes grow too long.  For the most part
-         * this only applies to "index" nodes (i.e., nodes where the values
-         * are much smaller than the keys).
+        /* Repeated compaction of tiny kvsets can make a node grow long
+         * and push the run-length based k-compaction far into the future.
+         * We address that here by looking for a run of kvsets with only
+         * a small number of keys.
          */
-        if (kvsets > runlen_max && cn_ns_vblks(&tn->tn_ns) < kvsets) {
+        if (kvsets > runlen_min + runlen_max) {
+            ulong keys_max = 32ul << 20;
+
             *mark = list_last_entry(head, typeof(*le), le_link);
             *action = CN_ACTION_COMPACT_KV;
             *rule = CN_RULE_INDEXF;
+            runlen = 0;
 
-            /* If the oldest kvset is larger than the next oldest kvset
-             * and there's not much garbage then start with the next
-             * oldest kvset, otherwise start with the oldest kvset.
-             */
-            le = list_prev_entry(*mark, le_link);
+            list_for_each_entry(le, head, le_link) {
+                const struct kvset_stats *stats = kvset_statsp(le->le_kvset);
 
-            if (kvset_get_kwlen(le->le_kvset) < kvset_get_kwlen((*mark)->le_kvset) &&
-                cn_ns_clen(&tn->tn_ns) * 100 > cn_ns_alen(&tn->tn_ns) * 85) {
+                if (stats->kst_keys > keys_max)
+                    break;
 
-                *rule = CN_RULE_INDEXP;
-                *mark = le;
-                kvsets--;
+                keys_max -= stats->kst_keys;
+                runlen++;
             }
 
-            return min_t(uint, kvsets, runlen_max);
+            if (runlen > runlen_min) {
+                if (runlen < kvsets) {
+                    *rule = CN_RULE_INDEXP;
+                    *mark = le;
+                }
+
+                return min_t(uint, runlen, runlen_max);
+            }
```

Comment:
This chunk of code worked find when key counts per node was small, but now that they can be very high it was doing way too much work.  Since adjusting the trigger up by adding runlen_min I haven't seen it trigger, which might mean the block to address long nodes (above) is handling the job.  Longer term, we might be able to just delete this block, will keep an eye on it.